test/gtest/uct/ib/test_devx.cc

```diff
@@ -48,4 +48,202 @@ UCS_TEST_P(test_devx, dbrec)
+
+    /* Now reuse the keys for different handles of different size */
+    const size_t REUSE_COUNT = 5;
+    for (size_t i = 0; i < REUSE_COUNT; ++i) {
+        memh[i] = create_memh(2048);
+        ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh[i]));
+        check_memh_export(memh[i], true);
+    }
+
+    ASSERT_EQ(MEMH_COUNT - REUSE_COUNT, ucs_list_length(&md()->umr.mkey_pool));
+
+    for (size_t i = 0; i < REUSE_COUNT; ++i) {
+        destroy_memh(memh[i]);
+    }
+
+    ASSERT_EQ(MEMH_COUNT, ucs_list_length(&md()->umr.mkey_pool));
+}
+
+UCS_TEST_P(test_devx_umr_mkey, import_mkey_no_umr)
+{
+    /* Export any mkey */
+    uct_ib_mlx5_devx_mem_t *exported_memh = create_memh(1024);
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), exported_memh));
+
+    /* Construct packed mkey without UMR tag */
+    uct_ib_md_packed_mkey_t packed_mkey = {
+        .lkey = exported_memh->exported_lkey & ~UCT_IB_MLX5_MKEY_TAG_UMR,
+        .vhca_id = md()->super.vhca_id
+    };
+
+    /* Import as non-UMR mkey */
+    uct_md_mem_attach_params_t params = {};
+    uct_ib_mlx5_devx_mem_t *imported_memh;
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_mem_attach(m_e->md(), &packed_mkey, &params,
+                                              (uct_mem_h *)&imported_memh));
+
+    /* UMR hash map remains uninitialized */
+    ASSERT_EQ(nullptr, md()->umr.mkey_hash);
+
+    destroy_memh(imported_memh);
+    destroy_memh(exported_memh);
+}
+
+UCS_TEST_P(test_devx_umr_mkey, import_mkey_umr)
+{
+    if (!(md()->flags & UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI)) {
+        UCS_TEST_SKIP_R("XGVMI is not supported");
+    }
+
+    /* Export UMR mkey */
+    uct_ib_mlx5_devx_mem_t *exported_memh = create_memh(1024);
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), exported_memh));
+    check_memh_export(exported_memh, true);
+
+    /* Construct packed mkey with UMR tag */
+    uct_ib_md_packed_mkey_t packed_mkey = {
+        .lkey = exported_memh->exported_lkey,
+        .vhca_id = md()->super.vhca_id
+    };
+
+    /* Import first UMR mkey */
+    uct_md_mem_attach_params_t params = {};
+    uct_ib_mlx5_devx_mem_t *imported_memh;
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_mem_attach(m_e->md(), &packed_mkey, &params,
+                                              (uct_mem_h *)&imported_memh));
+
+    check_umr_init(true, true);
+    ASSERT_EQ(1, kh_size(md()->umr.mkey_hash));
+
+    /* Second import with the same packed mkey, should reuse alias */
+    uct_ib_mlx5_devx_mem_t *imported_memh2;
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_mem_attach(m_e->md(), &packed_mkey, &params,
+                                              (uct_mem_h *)&imported_memh2));
+    ASSERT_NE(imported_memh, imported_memh2);
+    ASSERT_EQ(imported_memh->super.lkey, imported_memh2->super.lkey);
+    ASSERT_EQ(1, kh_size(md()->umr.mkey_hash));
+
+    destroy_memh(imported_memh);
+    destroy_memh(imported_memh2);
+    destroy_memh(exported_memh);
+}
+
 UCT_INSTANTIATE_IB_TEST_CASE(test_devx);
```

Comment:
pls move it before the new class, typically we initialize test classes right after their (and their tests) definitions

Reply:
sure